src/android/BrotherPrinter.java

```diff
@@ -581,4 +594,32 @@ public void onReceive(Context context, Intent intent) {
         });
     }
 
+    private void copyBinFile(String filename, String targetPath) {
+        AssetManager assetManager = cordova.getActivity().getAssets();
+        InputStream in = null;
+        OutputStream out = null;
+        String newFileName = null;
+        try {
+            in = assetManager.open(filename);
+
+            newFileName = targetPath + filename.substring(filename.lastIndexOf("/")+1);
+
+            out = new FileOutputStream(newFileName);
+
+            byte[] buffer = new byte[1024];
+            int read;
+            while ((read = in.read(buffer)) != -1) {
+                out.write(buffer, 0, read);
+            }
+            in.close();
```

Comment:
Should we use the java "try with resources" or do the closing in a "finally" block?

Reply:
Sorry, thought I'd replied to this already.

Yes, I agree with @arcadius. Either `finally` or `try-with-resources` would resolve the potential problem of making sure `close` runs if `read` or `write` throws an exception. Neither of them resolves the potential problem of either `close` throwing an exception of its own, but `try-with-resources` would at least guarantee that if both `read/write` and `close` throw both streams would try to close and you'd see the more important exception. (Sure, if `read` or `write` throws `flush` won't run, and does anybody care?). In my own Java 7 code I've used a construct that does essentially what Java 8 `try-with-exception` does, eating the `close` exception if both throw. So I'd recommend something like this:

    private void copyBinFile(String filename, String targetPath) {
        String newFileName = targetPath + filename.substring(filename.lastIndexOf("/")+1);
        AssetManager assetManager = cordova.getActivity().getAssets();
        try  (
                InputStream in = assetManager.open(filename);
                OutputStream out = new FileOutputStream(newFileName)
         ) {
            byte[] buffer = new byte[1024];
            int read;
            while ((read = in.read(buffer)) != -1) {
                out.write(buffer, 0, read);
            }
            out.flush();
        } catch (Exception e) {
            Log.e("Brother/SDKEvent", "Exception in copyFile() " + e.toString());
        }
    }